internal/osbuild/assembler_test.go

```diff
@@ -0,0 +1,230 @@
+					Size:        0,
+					Compression: "xz",
+				},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.tar","options":{"filename":"root.tar.xz","size":0,"compression":"xz"}}`),
+			},
+		},
+		{
+			name: "rawfs assembler empty",
+			assembler: Assembler{
+				Name:    "org.osbuild.rawfs",
+				Options: &RawFSAssemblerOptions{},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.rawfs","options":{"filename":"","root_fs_uuid":"00000000-0000-0000-0000-000000000000","size":0}}`),
+			},
+		},
+		{
+			name: "rawfs assembler full",
+			assembler: Assembler{
+				Name: "org.osbuild.rawfs",
+				Options: &RawFSAssemblerOptions{
+					Filename:           "filesystem.img",
+					RootFilesystemUUID: uuid.MustParse("76a22bf4-f153-4541-b6c7-0332c0dfaeac"),
+					Size:               2147483648,
+				},
+			},
+			args: args{
+				data: []byte(`{"name":"org.osbuild.rawfs","options":{"filename":"filesystem.img","root_fs_uuid":"76a22bf4-f153-4541-b6c7-0332c0dfaeac","size":2147483648}}`),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			assert := assert.New(t)
+			assembler := &tt.assembler
+			var gotAssembler Assembler
+			err := gotAssembler.UnmarshalJSON(tt.args.data)
+			assert.Equalf((err != nil), tt.wantErr, "Assembler.UnmarshalJSON() error = %v, wantErr %v", err, tt.wantErr)
+			if tt.wantErr {
+				return
+			}
+			gotBytes, err := json.Marshal(assembler)
+			assert.Nilf(err, "Could not marshal assembler: %v", err)
+			assert.Equalf(gotBytes, tt.args.data, "Expected `%v`, got `%v`", string(tt.args.data), string(gotBytes))
+			assert.Truef(reflect.DeepEqual(&gotAssembler, assembler),
+				"got {%v, %v}, expected {%v, %v}",
+				gotAssembler.Name,
+				gotAssembler.Options,
+				assembler.Name,
+				assembler.Options,
+			)
+		})
+	}
+}
+
+func TestNewQEMUAssembler(t *testing.T) {
+	options := &QEMUAssemblerOptions{
+		Format:   "qcow2",
+		Filename: "disk.qcow2",
+		Size:     2147483648,
+		PTUUID:   "0x14fc63d2",
+		PTType:   "mbr",
+		Partitions: []QEMUPartition{QEMUPartition{
+			Start:    2048,
+			Bootable: true,
+			Filesystem: QEMUFilesystem{
+				Type:       "ext4",
+				UUID:       "76a22bf4-f153-4541-b6c7-0332c0dfaeac",
+				Label:      "root",
+				Mountpoint: "/",
+			},
+		}},
+	}
+	expectedAssembler := &Assembler{
+		Name: "org.osbuild.qemu",
+		Options: &QEMUAssemblerOptions{
+			Format:   "qcow2",
+			Filename: "disk.qcow2",
+			Size:     2147483648,
+			PTUUID:   "0x14fc63d2",
+			PTType:   "mbr",
+			Partitions: []QEMUPartition{QEMUPartition{
+				Start:    2048,
+				Bootable: true,
+				Filesystem: QEMUFilesystem{
+					Type:       "ext4",
+					UUID:       "76a22bf4-f153-4541-b6c7-0332c0dfaeac",
+					Label:      "root",
+					Mountpoint: "/",
+				},
+			}},
+		},
+	}
+	assert.Equal(t, expectedAssembler, NewQEMUAssembler(options))
+}
+
+func TestNewTarAssembler(t *testing.T) {
+	options := NewTarAssemblerOptions("disk.tar.xz", 0)
+	options.SetCompression("xz")
```

Comment:
This should not be here. It is changing the expected behavior/meaning of this scenario. 

The constructor doesn't call `SetCompression` so you should assert that there is no compression by default and design a separate scenario which will assert that after calling SetCompression the options have been changed (but that is not related to the behavior of New).

Reply:
That's true, so I moved the use of `SetCompression` into a separate test.